nx/lib/nx/shape.ex

```diff
@@ -1218,6 +1218,69 @@ defmodule Nx.Shape do
     {List.to_tuple(result_shape), result_names}
   end
 
+  @doc """
+  Returns shape if valid and raises error if not.
+  """
+  def take_diagonal(tensor) do
+    case tensor.shape do
+      {len, breadth} ->
+        {len, breadth}
+
+      _bad_shape ->
+        raise ArgumentError, "take_diagonal/2 expects tensor of rank 2, got:\n#{inspect(tensor)}"
+    end
+  end
+
+  @doc """
+  Returns shape if valid and raises error if not.
+  """
+  def make_diagonal(tensor) do
+    case tensor.shape do
```

Comment:
Should I call `shape` instead? If a person passes a list instead of a tensor, it will fail with an `apply/3` error.

```suggestion
    case Nx.shape(tensor) do
```

Reply:
The functions in `Nx.Shape` do not receive the tensor, they always receive the shape, so you should change it do the same.

Back in the Nx module, you should do:

```elixir
def make_diagonal(tensor, opts \\ []) do
  tensor = to_tensor(tensor)
```

and then you can call `tensor.shape`, as it is guaranteed to be normalized.

Reply:
Will update `take_diagonal` and `make_diagonal`